### PR TITLE
fix(python): Set envvar for runfiles manifest, not runfiles dir, when using a manifest

### LINKED ISSUE #17675: Python wrapper sometimes puts manifest filename into RUNFILES_DIR
### Description of the bug:

py_binary rules come with a auto-generated [wrapper script](https://github.com/bazelbuild/bazel/blob/master/tools/python/python_bootstrap_template.txt) which automatically detects runfiles and sets either `RUNFILES_DIR` or `RUNFILES_MANIFEST_FILE`, depending on what is found first.

This script has a bug: if `progname.runfiles_manifest` file is missing, but `programe.runfiles/MANIFEST` is present, the script erroneously sets `RUNFILES_DIR` to manifest file location. This causes all sorts of failures down the line (for example when using bazel's `runfiles.bash` rules)

Code in question: https://github.com/bazelbuild/bazel/blob/4427a1fb0cf537da82e1ca4841e0cecda99e1107/tools/python/python_bootstrap_template.txt#L239
(compare to previous stanza)

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

To demonstrate it, we need any bazel repo which uses python runfiles lib. Bazel's own `rules_pkg` is one example. 
First, check out and run the test. It should work: 
```
git clone https://github.com/bazelbuild/rules_pkg.git
cd rules_pkg
bazel run //tests:stamp_test  # This works
```
Now naively package the runfiles dir and try to use it. Note we've got `stamp_test.runfiles/MANIFEST` (inside runfiles) but forgot `stamp_test.runfiles_manifest`:
```
rm -rf /tmp/work && mkdir /tmp/work && cp -r bazel-bin/tests/stamp_test bazel-bin/tests/stamp_test.runfiles /tmp/work && /tmp/work/stamp_test
...
NotADirectoryError: [Errno 20] Not a directory: '/tmp/work/stamp_test.runfiles/MANIFEST/rules_pkg/tests/stamped_zip.zip'

----------------------------------------------------------------------
Ran 2 tests in 0.002s

FAILED (errors=2)
```
observe how it treats `stamp_test.runfiles/MANIFEST` file as a directory and fails.

### Which operating system are you running Bazel on?

Various Ubuntu versions (18.04, 20.04, 22.04)

### What is the output of `bazel info release`?

This is present in HEAD and was introduced in 2018. Testcases above were tested on 6.0.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
https://github.com/bazelbuild/bazel.git
4427a1fb0cf537da82e1ca4841e0cecda99e1107
4427a1fb0cf537da82e1ca4841e0cecda99e1107
```

### Have you found anything relevant by searching the web?

Nothing. I am guessing very few people use custom bazel packaging scripts that only copy one file but not both.

### Any other information, logs, or outputs that you want to share?

We have a workaround which is to copy both files, and it probably is more correct that way, too. I am reporting this bug for the unlikely case someone else encounters it.

Workarounds for my test case above:
- Include both manifest files:
```
rm -rf /tmp/work && mkdir /tmp/work && cp -r bazel-bin/tests/stamp_test{,.runfiles,.runfiles_manifest} /tmp/work && /tmp/work/stamp_test
..
OK
```

- Do not include any manifest files
```
rm -rf /tmp/work && mkdir /tmp/work && cp -r bazel-bin/tests/stamp_test{,.runfiles} /tmp/work && rm -f /tmp/work/stamp_test.runfiles/MANIFEST && /tmp/work/stamp_test
..
OK
```

### PR DESCRIPTION
**Pull request by fmeum**
Fixes #17675

### CODE REVIEW

**Comment by fmeum** [comment]
@rickeylev Could you review?

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by rickeylev** [comment]
Can we add a test for this? In src/test/shell/integration/python_stub_test.sh, we can easily plug in a no-op interpreter that just prints env, then grep for the RUNFILES_MANIFEST_FILE. I think we just need to set `--build_runfile_manifests` when doing the build and invoke it using `bazel-bin` to make the right code paths get followed.

**Comment by fmeum** [comment]
@rickeylev Started working on the test, but I can't figure out how to build an input but not an output manifest. `--build_runfile_manifests` does seem to control both. I could of course delete one from the output base manually, but that feels hacky.

**Comment by rickeylev** [comment]
Hmm....yeah I don't see a way to generate only one and not the other. Maybe it's specific to certain platforms? Or maybe if...ah, what if it's running from a zip file? CreateModuleSpace() will unzip to a temp directory. The zip will contain MANIFEST, but the temp directory won't have .runfiles_manifest

In the original issue description, they manually copy the runfiles tree elsewhere and run, which is approximately the same as unzipping elsewhere.

I also think it'd be fine to just delete the .runfiles_manifest manually. It's effectively the same as the other two approaches, just with less copy+pasting of things around.

Let's just put a comment in the test about the case we're trying to reproduce; it'll look pretty funny deleting files from the output tree like that.

File: tools/python/python_bootstrap_template.txt [hunk 1]
```diff
@@ -238,7 +238,7 @@ def RunfilesEnvvar(module_space):
   # Look for the runfiles "input" manifest, argv[0] + ".runfiles/MANIFEST"
```
Comment (rickeylev) [inline tools/python/python_bootstrap_template.txt:238]:
```suggestion
  # Look for the runfiles "input" manifest, argv[0] + ".runfiles/MANIFEST"
  # Normally .runfiles_manifest and MANIFEST are both present, but
  # the former will be missing for zip-based builds or if someone copies
  # the runfiles tree elsewhere.
```

**Comment by kshyanashree** [comment]
@bazel-io fork 6.2.0

**Comment by fmeum** [comment]
@rickeylev I tried to get this to work but noticed that I can't come up with a scenario in which the change becomes relevant: In ZIP-based builds the incorrect branch isn't reached since `IsRunningFromZip()` is checked earlier. With `--nobuild_runfile_links` and a non-ZIP-based build, Python can't run as it needs the runfiles directory structure laid out. I would need a scenario in which the runfiles directory can only be found using the manifest file, which in turn is found via `argv[0]` - but is there a natural case in which this is the case?

**Comment by rickeylev** [comment]
Ah darn. Ok, well, lets just go back to the other ideas:
* Manually deleting the `{name}.runfiles_manifest` file
* The steps described in the issue: `cp -r bazel-bin/foo bazel-bin/foo.runfiles /tmp/work && /tmp/work/foo`

If none of _that_ works, well, lets just omit the tests, because we can't figure out a way to test it, which is fine. That line of code is obviously wrong.

**Comment by keertk** [comment]
Hi @fmeum @rickeylev we're planning to start creating RCs for 6.2.0 sometime in early May. Checking if we still want to include this?

**Comment by fmeum** [comment]
@rickeylev I couldn't figure out how to reproduce this situation in tests, even with these ideas. I added the returns comment and your suggestions.

**Review by rickeylev — APPROVED** [review]

**Comment by rickeylev** [comment]
re: LocalDiffAwarenessIntegrationTest failure -- no idea what that is or why it would fail. Looking at the test, I don't see any mention of Python in it. I'm guessing its flaky mac test. I pressed retry.

In any case, I'd say this is fine to begin importing. Our internal test infrastructure will handle flaky tests better.

**Comment by fmeum** [comment]
@bazel-io flag

**Pull request closed without merging**